### PR TITLE
executor: optimize count distinct with single column

### LINKED ISSUE #7539: optimize count distinct a single column individually
The implementation of the aggregate function `count(distinct xx)` calculates the value of the input column, encodes it to a `byte` slice, and visits a `map` with that `byte` slice to check whether the input is a distinct value.

It results in a poor performance, which takes **1 min 16.23 sec**:

```sql
TiDB(10.0.1.4:4000)> select count(distinct L_QUANTITY) from lineitem;
+----------------------------+
| count(distinct L_QUANTITY) |
+----------------------------+
|                         50 |
+----------------------------+
1 row in set (1 min 16.23 sec)
```

But we can specially handle the scenario that there is only one parameter in the `count` function, which might be easy with the help of `xxxSet` defined in **executor/aggfuncs/sets.go**.

And after the optimization, `count(distinct)` a single column should be at least as fast as `avg(distinct)`, which takes **56.60 sec**:
```sql
TiDB(10.0.1.4:4000)> select avg(distinct L_QUANTITY) from lineitem;
+--------------------------+
| avg(distinct L_QUANTITY) |
+--------------------------+
|                     25.5 |
+--------------------------+
1 row in set (56.60 sec)
```

### PR DESCRIPTION
**Pull request by SunRunAway**
### What problem does this PR solve? 

Resolves #7539

### What is changed and how it works?

Treat individually if `count distinct` has only 1 argument.

```
Count/count(distinct)/bigint(20)-12                     3.68ms ± 1%  1.86ms ± 1%  -49.49%  (p=0.008 n=5+5)
Count/count(distinct)/float-12                          3.62ms ± 2%  2.35ms ± 2%  -35.07%  (p=0.008 n=5+5)
Count/count(distinct)/double-12                         3.59ms ± 2%  2.34ms ± 2%  -34.72%  (p=0.008 n=5+5)
Count/count(distinct)/decimal(11,0)-12                  6.72ms ± 1%  5.49ms ± 1%  -18.28%  (p=0.008 n=5+5)
Count/count(distinct)/char(1)-12                        4.62ms ± 3%  2.56ms ± 4%  -44.66%  (p=0.008 n=5+5)
Count/count(distinct)/date-12                           4.18ms ± 1%  4.35ms ± 6%   +3.99%  (p=0.008 n=5+5)
Count/count(distinct)/time-12                           3.85ms ± 2%  1.80ms ± 4%  -53.37%  (p=0.008 n=5+5)
Count/count(distinct)/json-12                           4.05ms ± 1%  4.04ms ± 2%     ~     (p=0.841 n=5+5)
```

Before,
```mysql
MySQL [test]> select count(distinct id) from test;
+--------------------+
| count(distinct id) |
+--------------------+
|           10000000 |
+--------------------+
1 row in set (17.64 sec)
```

After,
```mysql
MySQL [test]> select count(distinct id) from test;
+--------------------+
| count(distinct id) |
+--------------------+
|           10000000 |
+--------------------+
1 row in set (10.85 sec)
```

### Check List 

Tests 

 - Unit test

Side effects

 - Increased code complexity

### CODE REVIEW

**Comment by zz-jason** [comment]
What's the end-to-end performance improvement?

File: executor/aggfuncs/builder.go [hunk 1]
```diff
@@ -105,6 +105,24 @@ func buildCount(aggFuncDesc *aggregation.AggFuncDesc, ordinal int) AggFunc {
 	// use countOriginalWithDistinct.
 	if aggFuncDesc.HasDistinct &&
 		(aggFuncDesc.Mode == aggregation.CompleteMode || aggFuncDesc.Mode == aggregation.Partial1Mode) {
+		if len(base.args) == 1 {
+			// optimize with single column
+			// TODO: because Time and JSON does not have `hashcode()` or similar method
+			// so they're in exception for now.
+			// TODO: add hashCode method for all evaluate types (Decimal, Time, Duration, JSON).
```
Comment (XuHuaiyu) [inline executor/aggfuncs/builder.go]: `Decimal` and `Duration` can be removed from this comment.

File: executor/aggfuncs/func_count_distinct.go [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+
+		p.partialResult4Count++
+	}
+
+	return nil
+}
+
+type partialResult4CountDistinctDecimal struct {
+	partialResult4Count
+	valSet set.StringSet
```
Comment (XuHuaiyu) [inline executor/aggfuncs/func_count_distinct.go]: Should we also update avgDistinct and sumDistinct to use StringSet?

File: executor/aggfuncs/aggfunc_test.go [hunk 3]
```diff
@@ -422,3 +422,95 @@ func (s *testSuite) testMultiArgsAggFunc(c *C, p multiArgsAggTest) {
...
+		b.Fatal(err)
+	}
+	finalFunc = aggfuncs.Build(s.ctx, desc, 0)
+	resultChk.Reset()
+	b.Run(fmt.Sprintf("%v(distinct)/%v", p.funcName, p.dataTypes), func(b *testing.B) {
+		s.baseBenchmarkAggFunc(b, finalFunc, input, resultChk)
+	})
+}
+
+func (s *testSuite) baseBenchmarkAggFunc(b *testing.B,
```
Comment (qw4990) [inline executor/aggfuncs/aggfunc_test.go]: Only benchmarks now, could you add some correct tests for these functions?

File: executor/aggfuncs/builder.go (hunk 1 above)
Comment (SunRunAway) [reply executor/aggfuncs/builder.go]: It's better to have a unified method name

File: executor/aggfuncs/aggfunc_test.go (hunk 3 above)
Comment (SunRunAway) [reply executor/aggfuncs/aggfunc_test.go]: It's already added. See https://github.com/pingcap/tidb/pull/15323/files#diff-9ec818ad7d98db209ac3a77e3d84e1e6R43-R56

File: executor/aggfuncs/func_count_distinct.go (hunk 2 above)
Comment (SunRunAway) [reply executor/aggfuncs/func_count_distinct.go]: They are already using `StringSet`.

File: executor/aggfuncs/func_count_distinct.go [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+		if isNull {
+			continue
+		}
+
+		if p.valSet.Exist(input) {
+			continue
+		}
+		p.valSet.Insert(input)
+
+		p.partialResult4Count++
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply executor/aggfuncs/func_count_distinct.go]: Updated.

File: executor/aggfuncs/builder.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (SunRunAway) [reply executor/aggfuncs/builder.go]: Added.

**Comment by SunRunAway** [comment]
> What's the end-to-end performance improvement?

Appended in PR description, PTAL.
@zz-jason @XuHuaiyu @qw4990

**Pull request merged**